blog/content/post/2020-06-30-trenchboot-grub-cbfs.md

```diff
@@ -0,0 +1,575 @@
+how it is related to TrenchBoot and GRUB? Answer is very simple. We prepare GRUB
+to support making operations on CBFS files (coreboot binaries) to create basis
+for possible LZ integration into coreboot's project. It means that in the
+feature, there will be possibility to have LZ already included in coreboot's
+binary. However, it is not reliable to take 64kB of space (LZ size), if we can
+compress it and take only 16kB. GRUB's part is essential here. First, it must
+read coreboot binary file layout. Second, it must decompress LZ, so it is
+prepared to be measured with DRTM. And that is exactly what we have introduced
+in this TrenchBoot project release. Now, GRUB have features to read SPI flash
+coreboot image and decompress any LZMA-compressed components.
+
+### GRUB features verification
+
+As every requirement, this one can also be validated by you. Before
+GRUB verification, first we need to preapre special coreboot image and update
+GRUB package to have additional features included.
+
+Hardware and firmware specification which we are using for test:
+
+- PC Engines apu2 platform
+- coreboot v4.12.0.2
+- NixOS operating system
+
+>We don't ensure valid test results if you don't have exactly the same
+configuration and environment as above.
+
+Procedure is divided into 3 sections:
+
+- coreboot image preparation
+- GRUB upgrade
+- feature verification
+
+#### coreboot image preparation
+
+Let's prepare coreboot image, which will include custom text file. This file
+should be compressed with LZMA. Later, GRUB should read content of coreboot
+image and decompress this message. To manipulate final coreboot binary file, one
+can use `cbfstool` utility. It is available with coreboot repository, in
+`coreboot/util/cbfstool` directory. We recommend to use
+[pcengines/coreboot](https://github.com/pcengines/coreboot/tree/develop/)
+repository for apu platforms. To prepare image with custom text file, follow the
+procedure below.
+
+1. Download latest coreboot release from [github.io](https://pcengines.github.io/)
+
+    ```
+    $ wget https://3mdeb.com/open-source-firmware/pcengines/apu2/apu2_v4.12.0.2.rom -O apu2_v4.12.0.2.rom
+    ```
+
+    > At time of writing this blog post, v4.12.0.2.rom was latest release
+
+2. Clone
+[pcengines/coreboot](https://github.com/pcengines/coreboot/tree/develop/)
+repository.
+
+    ```
+    $ git clone https://github.com/pcengines/coreboot -b develop
+    ```
+
+3. Enter `coreboot/util/cbfstool` directory and build `cbfstool`.
+
+    ```
+    $ cd coreboot/util/cbfstool
+    $ make
+    ```
+
+    If built was successful, there should be executable file `./cbfstool`
+    available.
+
+4. Create text file with any content you wish.
+
+    Only limitation here is to have long enough content, so LZMA will be done.
+    If file is too short (e.g. it contains only one sentence), most probably
+    LZMA won't execute on it and it won't be compressed. Name the file
```

Comment:
Reason for this should be explained, e.g.

```suggestion
    LZMA will produce a file bigger than the original (because of the compression header and dictionary) and `cbfstool` will choose the uncompressed form instead. Name the file
```

Reply:
Changed